### ISSUE TITLE
Fragments do not combine arrays correctly

### ISSUE THREAD

**Issue by chrbala** [dotansimha/graphql-code-generator#6149]
**Describe the bug**
Including multiple fragments that contain the same arrays results in the types getting combined at the array level. TS doesn't know how to merge these correctly. That is ``Array<TypeA> & Array<TypeB>`` is not equivalent to ``Array<TypeA & TypeB>`` because TS doesn't merge arrays.

**To Reproduce**
This repro has the bug fully reproducible.

https://github.com/chrbala/apollo-bug-repro/tree/codegen-merge-arrays

1. My GraphQL schema:

```graphql
type Item {
  string: String!
  number: Int!
}

type Query {
  items: [Item!]!
}
```

2. My GraphQL operations:

```graphql
query Page {
  ...String
  ...Number
}

fragment String on Query {
  items {
    string
  }
}

fragment Number on Query {
  items {
    number
  }
}
```

3. My `codegen.yml` config file:

```yml
overwrite: true
schema: "src/schema.js"
documents: "src/**/*.tsx"
generates:
  src/generated/graphql.tsx:
    plugins:
      - "typescript"
      - "typescript-operations"
      - "typescript-react-apollo"
```

**Expected behavior**
You should be able to access both the ``string`` and ``number`` properties on ``items``, but you can only access string. If you switch the order of ``...String`` and ``...Number``, then you can only access number. 

The generated fragment is
```ts
export type PageQuery = (
  { __typename?: 'Query' }
  & NumberFragment
  & StringFragment
);
```

But for it to work correctly in TS, it would need to be more like
```
export type PageQuery = (
  { __typename?: 'Query' }
  & {
    items: Array<NumberFragment['items'][0] & StringFragment['items'][0]>
  }
);
```

**Environment:**

- OS: OSX 10.14.6
- Codegen
    "@graphql-codegen/cli": "1.21.5",
    "@graphql-codegen/introspection": "1.18.2",
    "@graphql-codegen/typescript": "1.22.1",
    "@graphql-codegen/typescript-operations": "1.18.0",
    "@graphql-codegen/typescript-react-apollo": "2.2.5",
- NodeJS: v12.13.1

**Comment by dotansimha** [comment]
@n1ru4l what do you think?

**Comment by n1ru4l** [comment]
@dotansimha This is a bit tricky, as we generate types per fragment, but maybe there is some kind of merge TypeScript utility that can take two deep types and merge them?

```tsx
export type PageQuery = (
  { __typename?: 'Query' } 
  & DeepMerge<NumberFragment, StringFragment>
)
```

It seems like this would be possible via recursive conditional types: https://devblogs.microsoft.com/typescript/announcing-typescript-4-1-rc/#recursive-conditional-types

@chrbala If you could create a PR with a failing test case, that would be highly appreciated!

I can have a in-depth look at it this week!

**Comment by dotansimha** [comment]
Seems like it's possible to do with basic type inference (see https://www.typescriptlang.org/play?#code/C4TwDgpgBAglC8sBOSCGIA8BnYSCWAdgOYB8A3AFCiRQBCCyamBArgLYBGES5FV40ALLciEGCgwAZCADNgAZTwATCABooAJTxEAFguUQSDaXMUqoEAB7AIBJVkboMhGdygngAFQFGA-Ju09M2grGzsHcSZnAlckAN0vHyh-SKcPbxoAH3i9DMMoAC4oAggAN25KfhphJFElBhrRSIwYdVpeCgBjAHsCHCgbHCLGiHrEAG0KKCgAchx8YhnVKagARgAGCgBdMiA)

**Comment by n1ru4l** [comment]
Yes, this example is more close to the example provided by @chrbala : https://www.typescriptlang.org/play?ssl=10&ssc=3&pln=1&pc=1#code/C4TwDgpgBAglC8sBOSCGIA8BvKBnYSAlgHYDmAXHgSaVAL4B8A3AFCiRQBCCyamOxAK4BbAEYQklIWIn1mLNuGgBZCaQgwUGADIQAZsADKhACYQANFABKhUgAsjpiAx66Dxs1AgAPYBGImuLzoGCR6sm7AACpKLgD81rYOHtA+fgFBmnyhxOFIifbRsVAJWSGRMRwAZAUOlc5QUhAAbrJNrUisihyqSOomPL3qWRgwlpzyLADGAPbE+FB++JRDEAOIANosUFA4+ERklABER5bS4pJQAIz0LAC6TEA

**Comment by n1ru4l** [comment]
I created some test operations that should all cause this issue:

**Schema**
```graphql
type Query {
  user: User
}

type User {
  id: ID!
  name: String!
  friends: [User!]!
}


```

**Operations**
```graphql
query InlineFragmentQuery {
  user {
    ... on User {
      friends {
        id
      }
    }
    ... on User {
      friends {
        name
      }
    }
  }
}

fragment UserFriendsIdFragment on Query {
  user {
    friends {
      id
    }
  }
}

fragment UserFriendsNameFragment on Query {
  user {
    friends {
      name
    }
  }
}

query SpreadFragmentQuery {
  ...UserFriendsIdFragment
  ...UserFriendsNameFragment
}

query SpreadFragmentWithSelectionQuery {
  user {
    id
  }
  ...UserFriendsNameFragment
}

query InlineFragmentWithSpreadFragmentQuery {
  ... on Query {
    user {
      id
    }
  }
  ...UserFriendsNameFragment
}

query NestedQuery {
  user {
    id
    friends {
      friends {
        id
      }
    }
  }
  ... on User {
    friends {
      friends {
        name
      }
    }
  }
}

```

**Comment by n1ru4l** [comment]
Hey @chrbala, could you please explain your use case further? Deep merging fragment types seems to be a very complex problem.

In a framework such as relay, fragments are meant to describe data requirements of specific application components. Re-using fragments across multiple application components is considered an anti pattern

Here a small React example:

**UserFriendList.tx**

```tsx
import {  UserFriendsNameFragment } from "./generated";

// component only needs access to user names
export const UserFriendsList = (props: { data: UserFriendsNameFragment }) => {
  return ....
}
```

**UnfriendAllFriendsButton.tsx**

```tsx
import {  UserFriendsIdFragment } from "./generated";

// component only needs ids
export const UserFriendsList = (props: { data: UserFriendsIdFragment }) => {
  return ....
}
```

**App.tsx**

```tsx
import { useSpreadFragmentQuery } from "./generated";
import { UserFriendsList } from "./UserFriendList";
import { UnfriendAllFriendsButton } from "./UnfriendAllFriendsButton";

const App = () => {
  const data = useSpreadFragmentQuery();

  if (!data) {
    return null;
  }

  return (
    <>
      <UserFriendsList data={data} />
      <UnfriendAllFriendsButton data={data} />
    </>
  );
};
```


Of course, GraphQL is not only used in the context of building reusable UI components. Thus it would be great to know your use case and the reasons, why you want to use Fragments instead of just writing the selection, set inline.

A solution I already thought of would be to allow annotating a fragment spread with `@inlineFragment` (or a comment), to instruct graphql-codegen to inline the fragment type definition (similar to how inline fragments are handled today).

```graphql

query Page {
  ...String @inlineFragment @inlineFragment
  ...Number @inlineFragment
}

fragment String on Query {
  items {
    string
  }
}

fragment Number on Query {
  items {
    number
  }
}
```

---->>>>

```tsx
type PageQuery {
  __typename: "Query",
  items: Array<{ string: string; number: number }>
}
```

However, as I am typing this maybe we should instead make this the default behavior and always inline the fragment types within the Query documents.

**Comment by n1ru4l** [comment]
Hey @chrbala, I created https://github.com/dotansimha/graphql-code-generator/pull/6184 which changes the default behavior to generate more friendly types for all use-cases.

You can install the changes of the PR locally via 

```bash
yarn add @graphql-codegen/typescript-operations@1.18.3-alpha-33caf3be.0 @graphql-codegen/typescript-react-apollo@2.2.8-alpha-33caf3be.0 @graphql-codegen/typescript@1.22.4-alpha-33caf3be.0
```

(as new commits are pushed the list of packages is updated [over here](https://github.com/dotansimha/graphql-code-generator/pull/6184#issuecomment-867851741)

Please let me know whether this properly addresses your issues!

**Comment by chrbala** [comment]
My use case is similar to the Relay case actually. I'm using Apollo with colocated fragments next to components, as is done with Relay. Rather than having full data masking, I rely on static types to enforce which fields components can use.

This issue wasn't a hard blocker for me, since I was able to unblock by taking on some tech debt in adding fields to fragments that they don't belong in. So, I no longer have a live example to test against. However, it does look like this change fixes the issue for the example repo I provided. Thanks!

**Comment by strass** [comment]
I think I'm having a similar problem where I have a standard fragment that gets basic data that I want to extend on a case-by-case basis for specific pages.

```
fragment Basic on Foo {
  id
  label
  nested {
    edges {
      node {
        id
       label
      }
    }
  }
}

query Extended {
  allFoo {
    edges {
      node {
        ...Basic
       nested {
        id
        somethingNotInBasic
      }
      }
  }
}
```

If I try this, the fragment basically gets overwritten and I see `property label does not exist on type Nested` even though I would expect it to be there due to the merged fragment. Is this a correct assumption?

I tried using the pre-release that n1ru4l mentioned above but I get this error when I try to run codegen:
```
 Found 1 error

  ✖ src/graphql/
    TypeError: Invalid state.
        at mergeSelectionSets (/Users/strsszd/projects/Espresso/frontend/node_modules/@graphql-codegen/typescript-operations/node_modules/@graphql-codegen/visitor-plugin-common/index.cjs.js:
450:19)
        at mergeSelectionSets (/Users/strsszd/projects/Espresso/frontend/node_modules/@graphql-codegen/typescript-operations/node_modules/@graphql-codegen/visitor-plugin-common/index.cjs.js:
459:35)
        at mergeSelectionSets (/Users/strsszd/projects/Espresso/frontend/node_modules/@graphql-codegen/typescript-operations/node_modules/@graphql-codegen/visitor-plugin-common/index.cjs.js:
459:35)
        at mergeSelectionSets (/Users/strsszd/projects/Espresso/frontend/node_modules/@graphql-codegen/typescript-operations/node_modules/@graphql-codegen/visitor-plugin-common/index.cjs.js:
459:35)
        at mergeSelectionSets (/Users/strsszd/projects/Espresso/frontend/node_modules/@graphql-codegen/typescript-operations/node_modules/@graphql-codegen/visitor-plugin-common/index.cjs.js:
459:35)
        at mergeSelectionSets (/Users/strsszd/projects/Espresso/frontend/node_modules/@graphql-codegen/typescript-operations/node_modules/@graphql-codegen/visitor-plugin-common/index.cjs.js:
459:35)
        at mergeSelectionSets (/Users/strsszd/projects/Espresso/frontend/node_modules/@graphql-codegen/typescript-operations/node_modules/@graphql-codegen/visitor-plugin-common/index.cjs.js:
459:35)
        at mergeSelectionSets (/Users/strsszd/projects/Espresso/frontend/node_modules/@graphql-codegen/typescript-operations/node_modules/@graphql-codegen/visitor-plugin-common/index.cjs.js:
459:35)
        at mergeSelectionSets (/Users/strsszd/projects/Espresso/frontend/node_modules/@graphql-codegen/typescript-operations/node_modules/@graphql-codegen/visitor-plugin-common/index.cjs.js:
459:35)
        at SelectionSetToObject.buildSelectionSetString (/Users/strsszd/projects/Espresso/frontend/node_modules/@graphql-codegen/typescript-operations/node_modules/@graphql-codegen/visitor-p
lugin-common/index.cjs.js:2698:51)
    TypeError: Invalid state.
        at mergeSelectionSets (/Users/strsszd/projects/Espresso/frontend/node_modules/@graphql-codegen/typescript-operations/node_modules/@graphql-codegen/visitor-plugin-common/index.cjs.js:
450:19)
        at mergeSelectionSets (/Users/strsszd/projects/Espresso/frontend/node_modules/@graphql-codegen/typescript-operations/node_modules/@graphql-codegen/visitor-plugin-common/index.cjs.js:
459:35)
        at mergeSelectionSets (/Users/strsszd/projects/Espresso/frontend/node_modules/@graphql-codegen/typescript-operations/node_modules/@graphql-codegen/visitor-plugin-common/index.cjs.js:
459:35)
        at mergeSelectionSets (/Users/strsszd/projects/Espresso/frontend/node_modules/@graphql-codegen/typescript-operations/node_modules/@graphql-codegen/visitor-plugin-common/index.cjs.js:
459:35)
        at mergeSelectionSets (/Users/strsszd/projects/Espresso/frontend/node_modules/@graphql-codegen/typescript-operations/node_modules/@graphql-codegen/visitor-plugin-common/index.cjs.js:
459:35)
        at mergeSelectionSets (/Users/strsszd/projects/Espresso/frontend/node_modules/@graphql-codegen/typescript-operations/node_modules/@graphql-codegen/visitor-plugin-common/index.cjs.js:
459:35)
        at mergeSelectionSets (/Users/strsszd/projects/Espresso/frontend/node_modules/@graphql-codegen/typescript-operations/node_modules/@graphql-codegen/visitor-plugin-common/index.cjs.js:
459:35)
        at mergeSelectionSets (/Users/strsszd/projects/Espresso/frontend/node_modules/@graphql-codegen/typescript-operations/node_modules/@graphql-codegen/visitor-plugin-common/index.cjs.js:
459:35)
        at mergeSelectionSets (/Users/strsszd/projects/Espresso/frontend/node_modules/@graphql-codegen/typescript-operations/node_modules/@graphql-codegen/visitor-plugin-common/index.cjs.js:
459:35)
        at SelectionSetToObject.buildSelectionSetString (/Users/strsszd/projects/Espresso/frontend/node_modules/@graphql-codegen/typescript-operations/node_modules/@graphql-codegen/visitor-p
lugin-common/index.cjs.js:2698:51)
```

This isn't technically about array fragments so let me know if I should make a new issue.

**Comment by n1ru4l** [comment]
@strass can you provide a minimal reproduction code-sandbox or pr with failing test?

**Comment by strass** [comment]
Our schema isn't really conducive towards a MWE... Are there any sandboxes out there that I could build on?

**Comment by chrbala** [comment]
Can I suggest:
git: https://github.com/apollographql/react-apollo-error-template
browser: https://codesandbox.io/s/github/apollographql/react-apollo-error-template

You can also use [my example](https://github.com/chrbala/apollo-bug-repro/tree/codegen-merge-arrays) for schema mocking.

**Comment by strass** [comment]
I wanted to use a server template so I could regen gql files on the sandbox itself but it's a little finnicky to use...

https://codesandbox.io/s/gql-test-qktgl?file=/src/index.tsx

I think you might have to open a new terminal window and run `yarn start` manually

I would expect the RECURSIVE query to be able to merge the spread fragment with an inline query, but it ends up only using the inline query and can't access the fields specified via the Basic fragment
<img width="1213" alt="Screen Shot 2021-07-13 at 11 50 44 AM" src="https://user-images.githubusercontent.com/134242/125508549-8727d92d-f4e2-4d03-a96d-08f2ff262ceb.png">

**Comment by n1ru4l** [comment]
@strass The code sandbox you shared showcases the same issue that OP had 😅 This issue should be addressed by https://github.com/dotansimha/graphql-code-generator/pull/6184 (you can try out the packages within the PR)

The issue from your comment here: https://github.com/dotansimha/graphql-code-generator/issues/6149#issuecomment-879216617 seems like an issue when generating types, that should not happen... It would be great if you could share a reproduction for it

**Comment by n1ru4l** [comment]
@strass I tried to reproduce this error over here: https://github.com/dotansimha/graphql-code-generator/pull/6317

Maybe you could alter the test, so it fails?

**Comment by strass** [comment]
The PR title talks about arrays specifically so I wanted to make sure to give an example that wasn't array-based.

I'll take a look at what you've prepared. Thanks :)

**Comment by strass** [comment]
> @strass I tried to reproduce this error over here: #6317
> 
> Maybe you could alter the test, so it fails?

OK I made a small edit to the test and commented it in that PR (tried to push it back up but didnt have access). It runs successfully within the repo, but is failing in my development environment. I'm currently on:

>   "@graphql-codegen/cli": "^1.21.7",
>     "@graphql-codegen/fragment-matcher": "^2.0.1",
>     "@graphql-codegen/introspection": "^1.18.2",
>     "@graphql-codegen/near-operation-file-preset": "^1.18.5",
>     "@graphql-codegen/plugin-helpers": "^1.18.8",
>     "@graphql-codegen/typescript": "^1.23.0",
>     "@graphql-codegen/typescript-graphql-files-modules": "^1.18.1",
>     "@graphql-codegen/typescript-operations": "^1.18.4",
>     "@graphql-codegen/typescript-react-apollo": "^2.3.1",

I'll try updating with `npx npm-check-updates '/@graphql-codegen/' -u` and report back

**Comment by strass** [comment]
It's similar output but my fragment isn't being inlined

```
    Snapshot name: `graphql tests reproduction https://github.com/dotansimha/graphql-code-generator/issues/6149#issuecomment-879320765 1`

    - Snapshot  - 2
    + Received  + 2

    @@ -7,21 +7,21 @@
          { __typename?: 'FooConnection' }
          & { edges: Array<(
            { __typename?: 'FooEdge' }
            & { node: (
              { __typename?: 'Foo' }
    -         & Pick<Foo, 'id' | 'label'>
              & { nested: Maybe<(
                { __typename?: 'FooConnection' }
                & { edges: Array<(
                  { __typename?: 'FooEdge' }
                  & { node: (
                    { __typename?: 'Foo' }
    -               & Pick<Foo, 'id' | 'somethingNotInBasic' | 'label' | 'somethingNotInQuery'>
    +               & Pick<Foo, 'id' | 'somethingNotInBasic'>
                  ) }
                )> }
              )> }
    +         & BasicFragment
            ) }
          )> }
        )> }
      );
```

**Comment by dotansimha** [comment]
Available in `@graphql-codegen/typescript-operations@2.0.0`

**Comment by claycoleman** [comment]
Hi @dotansimha and @n1ru4l , I'm also running into errors related to this. It seems that if you request an array of interfaces field multiple times, spreading fragments on them leads to invalid state. This could commonly happen when a set of subcomponents define fragments on the same type and each request different fields on an interface array. I have a reproducible repo here: https://codesandbox.io/s/gql-test-forked-j275t?file=/src/index.tsx

You can pop a terminal and run `yarn codegen` to see the error. 

here's the schema and query:

**Schema**
```schema
interface IItem {
  sharedField: String
}

type ItemA implements IItem {
  sharedField: String
  itemAOnlyField: String
}

type ItemB implements IItem {
  sharedField: String
  itemBOnlyField: String 
}

type Parent { 
  parentField: String
  children: [IItem!]!
}

type Query {
  parent: Parent!
}
```

**Repro Query**
```query
query Page {
    parent {
      parentField
      children {
        ... on ItemA {
          itemAOnlyField
        }
      }
      children {
        ... on ItemA {
          itemAOnlyField
        }
      }
    }
  }
```

My real use case looks something closer to the below `gql` operations, but after playing around I determined that it also happens if you request the interface array multiple times even with the same fragment type (seen above as `children { ... on ItemA { itemAOnlyField } }`)

**Real Query**
```query
  # defined in FirstComponent.tsx
  fragment FirstComponent_parent on Parent {
    parentField
    children {
      sharedField
      ... on ItemA {
        itemAOnlyField
      }
    }
  }

  # defined in SecondComponent.tsx
  fragment SecondComponent_parent on Parent {
    parentField
    children {
      sharedField
      ... on ItemB {
        itemBOnlyField
      }
    }
  }

  # defined in ParentComponent.tsx
  query ParentComponentQuery {
    parent {
      parentField
      ...FirstComponent_parent
      ...SecondComponent_parent
    }
  }
```

The error I get when trying to compile: 


```
Found 1 error

  ✖ src/generated/graphql.tsx
    TypeError: Invalid state.
        at mergeSelectionSets (/sandbox/node_modules/@graphql-codegen/visitor-plugin-com
mon/index.cjs.js:447:19)
        at SelectionSetToObject.buildSelectionSetString (/sandbox/node_modules/@graphql-
codegen/visitor-plugin-common/index.cjs.js:2671:25)
        at /sandbox/node_modules/@graphql-codegen/visitor-plugin-common/index.cjs.js:262
1:41
        at Array.reduce (<anonymous>)
        at SelectionSetToObject._buildGroupedSelections (/sandbox/node_modules/@graphql-
codegen/visitor-plugin-common/index.cjs.js:2611:80)
        at SelectionSetToObject.transformSelectionSet (/sandbox/node_modules/@graphql-co
degen/visitor-plugin-common/index.cjs.js:2735:30)
        at SelectionSetToObject.buildSelectionSetString (/sandbox/node_modules/@graphql-
codegen/visitor-plugin-common/index.cjs.js:2685:89)
        at /sandbox/node_modules/@graphql-codegen/visitor-plugin-common/index.cjs.js:262
1:41
        at Array.reduce (<anonymous>)
        at SelectionSetToObject._buildGroupedSelections (/sandbox/node_modules/@graphql-
codegen/visitor-plugin-common/index.cjs.js:2611:80)
    TypeError: Invalid state.
        at mergeSelectionSets (/sandbox/node_modules/@graphql-codegen/visitor-plugin-com
mon/index.cjs.js:447:19)
        at SelectionSetToObject.buildSelectionSetString (/sandbox/node_modules/@graphql-
codegen/visitor-plugin-common/index.cjs.js:2671:25)
        at /sandbox/node_modules/@graphql-codegen/visitor-plugin-common/index.cjs.js:262
1:41
        at Array.reduce (<anonymous>)
        at SelectionSetToObject._buildGroupedSelections (/sandbox/node_modules/@graphql-
codegen/visitor-plugin-common/index.cjs.js:2611:80)
        at SelectionSetToObject.transformSelectionSet (/sandbox/node_modules/@graphql-co
degen/visitor-plugin-common/index.cjs.js:2735:30)
        at SelectionSetToObject.buildSelectionSetString (/sandbox/node_modules/@graphql-
codegen/visitor-plugin-common/index.cjs.js:2685:89)
        at /sandbox/node_modules/@graphql-codegen/visitor-plugin-common/index.cjs.js:262
1:41
        at Array.reduce (<anonymous>)
        at SelectionSetToObject._buildGroupedSelections (/sandbox/node_modules/@graphql-
codegen/visitor-plugin-common/index.cjs.js:2611:80)
```

Thanks for your help on this!
